source/02_algo/distributed_rl_eng.rst

```diff
@@ -0,0 +1,122 @@
+
+
+Depending on the algorithm and environment, some extended auxiliary modules will be added. For example, most off-policy algorithms will require a data queue (Replay Buffer) to store training data, and the ENV model will be learning for model-based RL-related algorithms. For the relevant training modules of environmental dynamics and algorithms that require a large number of self-play, a centralized Coordinator is also required to control and coordinate various components (such as dynamically communicating with both sides of the game).
+
+From a system perspective, it is necessary to allow sufficient parallel scalability for similar modules in the entire training program. For example, the number of interacting environments can be increased according to the availability of resources (utilize more CPUs), or the throughput of the training side can be increased according to the number of parallel devices (generally, having more GPUs). For different modules, it is hoped that all modules can be executed asynchronously as much as possible, and the cost of various communication methods (network communication, database, file system) in module time is reduced. But in general, the theoretical upper limit of the efficiency optimization of a system is achieved when the Learner can continuously train efficiently without waiting; that is, when the Learner completes one training iteration efficiently, the data for the next training iteration is already available. 
+
+From the algorithmic point of view, it is hoped to reduce the algorithm’s requirements for data throughput (such as tolerating older and more off-policy data) while ensuring the convergence of the algorithm, and improving the efficiency of data exploration and the utilization of collected data (for example, modify the data sampling method, or combine some research related to data-efficiency in RL). So that it provides more space and possibilities for system design.
+
+To sum up, distributed reinforcement learning is a more comprehensive research subfield, which requires mutual perception and coordination of deep reinforcement learning algorithm + distributed system design.
+
+
+Research Direction
+---------
+
+System
+~~~~~~
+
+Overall Architecture
+^^^^^^^^^^
+For common decision problems, the two most commonly used distributed architectures are IMPALA [3]_ and SEED RL [4]_.
+
+.. image:: ./images/impala.png
+  :align: center
+  
+- The former is the classic Actor-Learner mode; that is, the data collection and training sides are entirely separated, and the latest neural network model is regularly passed from the Learner to the Actor, and the observations collected by the Actor are sent to the Learner after collecting a certain amount of data (i.e. observations). If there are multiple Learners, they also periodically synchronize their gradients to update the neural network (i.e. the data-parallel model in distributed deep learning).
```

Comment:
the observations collected -> the data collected
a certain amount of data (i.e. observations) -> a certain amount of dataframes (i.e. observations)